### ISSUE TITLE
Messages from Discord to Slack duplicated on Discord side

### ISSUE THREAD

**Issue by qaisjp** [42wim/matterbridge#1039]
**Describe the bug**  
Message is duplicated in Discord because of bot message in Slack

**To Reproduce**

1. Set up simple test discord <-> slack bridge
2. Send message in Discord
3. Message is bridged to Slack, which is bridged back to Discord
4. The message you sent in Discord is duplicated in Discord

**Expected behavior**

Message should not be duplicated.

**Screenshots/debug logs**
Use logs from running `matterbridge -debug` if possible.

_Discord_

![image](https://user-images.githubusercontent.com/923242/76846912-2ae07880-6839-11ea-9d3e-e2827fefe9bc.png)

_Slack_

Appears as expected on Slack (with my photo and name).

_Logs_

<details><summary>click here</summary>

```
➜ ./matterbridge -debug
[0000]  INFO main:         Enabling debug logging.
[0000]  INFO main:         Running version 1.16.6-dev
[0000]  INFO main:         WARNING: THIS IS A DEVELOPMENT VERSION. Things may break.
[0000] DEBUG discord:      Configuring Discord Incoming Webhook
[0000]  INFO router:       Parsing gateway mta.test
[0000]  INFO router:       Starting bridge: slack.mta
[0000]  INFO slack:        Connecting using token
[0000]  INFO slack:        slack.mta: joining test (ID: testslack.mta)
[0000] DEBUG slack:        Choosing token based receiving
[0000] DEBUG slack:        == Receiving event &slack.ConnectingEvent{Attempt:1, ConnectionCount:0}
[0000] DEBUG slack:        Unhandled incoming event: *slack.ConnectingEvent
[0000]  INFO router:       Starting bridge: discord.mta
[0000]  INFO discord:      Connecting
[0000]  INFO discord:      Connecting using webhookurl (for posting) and token
[0000]  INFO discord:      Connection succeeded
[0000] DEBUG slack:        == Receiving event &slack.ConnectedEvent{ConnectionCount:0, Info:(*slack.Info)(0xc00040a000)}
[0001] DEBUG slack:        Start listening for Slack messages
[0001] DEBUG discord:      Verifying PermissionManageWebhooks for testdiscord.mta with ID 590172620440469505
[0001]  INFO discord:      Can manage webhooks; will edit channel for global webhook on send
[0001] DEBUG slack:        getting 33 users
[0002]  INFO discord:      discord.mta: joining test (ID: testdiscord.mta)
[0002]  INFO main:         Gateway(s) started succesfully. Now relaying messages
[0002] DEBUG discord:      == Receiving event &discordgo.Message{ID:"689415941343150083", ChannelID:"590172620440469505", GuildID:"278474088903606273", Content:"PLEASE FVV", Timestamp:"2020-03-17T10:12:49.145000+00:00", EditedTimestamp:"", MentionRoles:[]string{}, Tts:false, MentionEveryone:false, Author:(*discordgo.User)(0xc000574a80), Attachments:[]*discordgo.MessageAttachment{}, Embeds:[]*discordgo.MessageEmbed{}, Mentions:[]*discordgo.User{}, Reactions:[]*discordgo.MessageReactions(nil), Pinned:false, Type:0, WebhookID:"", Member:(*discordgo.Member)(0xc000576460), MentionChannels:[]*discordgo.Channel(nil), Activity:(*discordgo.MessageActivity)(nil), Application:(*discordgo.MessageApplication)(nil), MessageReference:(*discordgo.MessageReference)(nil), Flags:0}
[0002] DEBUG discord:      <= Sending message from qaisjp on discord.mta to gateway
[0002] DEBUG discord:      <= Message is config.Message{Text:"PLEASE FVV", Channel:"test", Username:"qaisjp", UserID:"83386293446246400", Avatar:"https://cdn.discordapp.com/avatars/83386293446246400/a525dac41568459ca38bafd903af8548.jpg", Account:"discord.mta", Event:"", Protocol:"", Gateway:"", ParentID:"", Timestamp:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}, ID:"689415941343150083", Extra:map[string][]interface {}(nil)}
[0002] DEBUG gateway:      => Sending config.Message{Text:"PLEASE FVV", Channel:"test", Username:"qaisjp", UserID:"83386293446246400", Avatar:"https://cdn.discordapp.com/avatars/83386293446246400/a525dac41568459ca38bafd903af8548.jpg", Account:"discord.mta", Event:"", Protocol:"discord", Gateway:"mta.test", ParentID:"", Timestamp:time.Time{wall:0xbf9446885119f1e8, ext:2615190372, loc:(*time.Location)(0x3c76020)}, ID:"689415941343150083", Extra:map[string][]interface {}(nil)} from discord.mta (test) to slack.mta (test)
[0002] DEBUG slack:        => Receiving config.Message{Text:"PLEASE FVV", Channel:"test", Username:"qaisjp", UserID:"83386293446246400", Avatar:"https://cdn.discordapp.com/avatars/83386293446246400/a525dac41568459ca38bafd903af8548.jpg", Account:"discord.mta", Event:"", Protocol:"discord", Gateway:"mta.test", ParentID:"", Timestamp:time.Time{wall:0xbf9446885119f1e8, ext:2615190372, loc:(*time.Location)(0x3c76020)}, ID:"", Extra:map[string][]interface {}(nil)}
[0002] DEBUG slack:        == Receiving event &slack.MessageEvent{Msg:slack.Msg{ClientMsgID:"", Type:"message", Channel:"C4Z7JEUC8", User:"", Text:"PLEASE FVV", Timestamp:"1584439969.005400", ThreadTimestamp:"", IsStarred:false, PinnedTo:[]string(nil), Attachments:[]slack.Attachment(nil), Edited:(*slack.Edited)(nil), LastRead:"", Subscribed:false, UnreadCount:0, SubType:"bot_message", Hidden:false, DeletedTimestamp:"", EventTimestamp:"1584439969.005400", BotID:"BF5S2HSTG", Username:"qaisjp", Icons:(*slack.Icon)(0xc0003e7880), Inviter:"", Topic:"", Purpose:"", Name:"", OldName:"", Members:[]string(nil), ReplyCount:0, Replies:[]slack.Reply(nil), ParentUserId:"", Files:[]slack.File(nil), Upload:false, Comment:(*slack.Comment)(nil), ItemType:"", ReplyTo:0, Team:"T216B1U03", Reactions:[]slack.ItemReaction(nil), ResponseType:"", ReplaceOriginal:false, DeleteOriginal:false, Blocks:slack.Blocks{BlockSet:[]slack.Block(nil)}}, SubMessage:(*slack.Msg)(nil), PreviousMessage:(*slack.Msg)(nil)}
[0002] DEBUG gateway:      mID slack.mta: 1584439969.005400
[0002] DEBUG slack:        Found bot &slack.Bot{ID:"BF5S2HSTG", Name:"Chat Bridge", Deleted:false, UserID:"UF3N0SGJU", AppID:"AF3RWLXBK", Updated:1546356234, Icons:slack.Icons{Image36:"https://a.slack-edge.com/80588/img/plugins/app/bot_36.png", Image48:"https://a.slack-edge.com/80588/img/plugins/app/bot_48.png", Image72:"https://a.slack-edge.com/80588/img/plugins/app/service_72.png"}}
[0002] DEBUG slack:        <= Sending message from qaisjp on slack.mta to gateway
[0002] DEBUG slack:        GetUserInfo failed for BF5S2HSTG: user_not_found
[0002] DEBUG slack:        <= Message is &config.Message{Text:"PLEASE FVV", Channel:"test", Username:"qaisjp", UserID:"BF5S2HSTG", Avatar:"", Account:"slack.mta", Event:"", Protocol:"slack", Gateway:"", ParentID:"", Timestamp:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}, ID:"1584439969.005400", Extra:map[string][]interface {}{}}
[0002] DEBUG gateway:      => Sending config.Message{Text:"PLEASE FVV", Channel:"test", Username:"qaisjp", UserID:"BF5S2HSTG", Avatar:"", Account:"slack.mta", Event:"", Protocol:"slack", Gateway:"mta.test", ParentID:"", Timestamp:time.Time{wall:0xbf94468866708f98, ext:2973192208, loc:(*time.Location)(0x3c76020)}, ID:"1584439969.005400", Extra:map[string][]interface {}{}} from slack.mta (test) to discord.mta (test)
[0002] DEBUG discord:      => Receiving config.Message{Text:"PLEASE FVV", Channel:"test", Username:"qaisjp", UserID:"BF5S2HSTG", Avatar:"", Account:"slack.mta", Event:"", Protocol:"slack", Gateway:"mta.test", ParentID:"", Timestamp:time.Time{wall:0xbf94468866708f98, ext:2973192208, loc:(*time.Location)(0x3c76020)}, ID:"", Extra:map[string][]interface {}{}}
[0002] DEBUG discord:      Broadcasting using Webhook
[0002] DEBUG discord:      Setting webhook channel to "test"
[0003] DEBUG discord:      Processing webhook sending for message config.Message{Text:"PLEASE FVV", Channel:"test", Username:"qaisjp", UserID:"BF5S2HSTG", Avatar:"", Account:"slack.mta", Event:"", Protocol:"slack", Gateway:"mta.test", ParentID:"", Timestamp:time.Time{wall:0xbf94468866708f98, ext:2973192208, loc:(*time.Location)(0x3c76020)}, ID:"", Extra:map[string][]interface {}{}}
[0003] DEBUG gateway:      mID discord.mta: 689415944501461043
```

</details>

**Environment (please complete the following information):**
 - OS: [e.g. linux]
 - Matterbridge version: output of  `matterbridge -version`
 - If self compiled: output of `git rev-parse HEAD`

_Problem happens on my test instance_

6b4b19194ec40c79569bc692a6557e78f338ded9 + #1036 (needed to make it compile), macOS

_Does not happen on my production instance_

c0be3e585a5ef6c6dc4d517e891fd10bc053c195 + few tiny patches, Linux

**Additional context**
Please add your configuration file  (be sure to exclude or anonymize private data (tokens/passwords))

Related to #219 

```toml
[discord]

[discord.mta]
Token="bot-token"
Server="278474088903606273"
RemoteNickFormat="{NICK}"
WebhookURL="https://discordapp.com/api/webhooks/684875846770753568/token"

[slack]
[slack.mta]
Token="xoxb-xxxxx-yyyy-zzzz"
RemoteNickFormat="{NICK}"


[[gateway]]
name="mta.test"
enable=true

[[gateway.inout]]
account="slack.mta"
channel="test"

[[gateway.inout]]
account="discord.mta"
channel="test"
```

**Comment by qaisjp** [comment]
Could be related to #1018. Doing some debugging...

With this patch on [this file](https://github.com/42wim/matterbridge/blob/6b4b19194ec40c79569bc692a6557e78f338ded9/bridge/slack/handlers.go)

```diff
diff --git a/bridge/slack/handlers.go b/bridge/slack/handlers.go
index 10ed0ae5..04d30fed 100644
--- a/bridge/slack/handlers.go
+++ b/bridge/slack/handlers.go
@@ -125,6 +125,10 @@ func (b *Bslack) skipMessageEvent(ev *slack.MessageEvent) bool {
        }

        // Skip any messages that we made ourselves or from 'slackbot' (see #527).
+       fmt.Printf("Comparing ev.Username[%#v] == sSlackBotUser[%#v]\n", ev.Username, sSlackBotUser)
+       fmt.Printf("Comparing (b.rtm!=nil)[%#v] && ev.Username[%#v] == b.si.User.Name[%#v]\n", b.rtm != nil, ev.Username, b.si.User.Name)
+       fmt.Printf("Comparing attachment stuff. Result %#v\n", (len(ev.Attachments) > 0 && ev.Attachments[0].CallbackID == "matterbridge_"+b.uuid))
+       fmt.Printf("Slack event: %#v\n", ev)
        if ev.Username == sSlackBotUser ||
                (b.rtm != nil && ev.Username == b.si.User.Name) ||
                (len(ev.Attachments) > 0 && ev.Attachments[0].CallbackID == "matterbridge_"+b.uuid) {
```

This output is given:
```
➜ ./matterbridge
[2020-03-17T10:52:27Z]  INFO main:         Running version 1.16.6-dev
[2020-03-17T10:52:27Z]  INFO main:         WARNING: THIS IS A DEVELOPMENT VERSION. Things may break.
[2020-03-17T10:52:27Z]  INFO router:       Parsing gateway mta.test
[2020-03-17T10:52:27Z]  INFO router:       Starting bridge: discord.mta
[2020-03-17T10:52:27Z]  INFO discord:      Connecting
[2020-03-17T10:52:27Z]  INFO discord:      Connecting using webhookurl (for posting) and token
[2020-03-17T10:52:27Z]  INFO discord:      Connection succeeded
[2020-03-17T10:52:28Z]  INFO discord:      Can manage webhooks; will edit channel for global webhook on send
[2020-03-17T10:52:29Z]  INFO discord:      discord.mta: joining test (ID: testdiscord.mta)
[2020-03-17T10:52:29Z]  INFO router:       Starting bridge: slack.mta
[2020-03-17T10:52:29Z]  INFO slack:        Connecting using token
[2020-03-17T10:52:29Z]  INFO slack:        slack.mta: joining test (ID: testslack.mta)
[2020-03-17T10:52:29Z]  INFO main:         Gateway(s) started succesfully. Now relaying messages
Comparing ev.Username["qaisjp"] == sSlackBotUser["slackbot"]
Comparing (b.rtm!=nil)[true] && ev.Username["qaisjp"] == b.si.User.Name["chat-bridge"]
Comparing attachment stuff. Result false
Slack event: &slack.MessageEvent{Msg:slack.Msg{ClientMsgID:"", Type:"message", Channel:"C4Z7JEUC8", User:"", Text:"yes", Timestamp:"1584442356.006000", ThreadTimestamp:"", IsStarred:false, PinnedTo:[]string(nil), Attachments:[]slack.Attachment(nil), Edited:(*slack.Edited)(nil), LastRead:"", Subscribed:false, UnreadCount:0, SubType:"bot_message", Hidden:false, DeletedTimestamp:"", EventTimestamp:"1584442356.006000", BotID:"BF5S2HSTG", Username:"qaisjp", Icons:(*slack.Icon)(0xc000475420), Inviter:"", Topic:"", Purpose:"", Name:"", OldName:"", Members:[]string(nil), ReplyCount:0, Replies:[]slack.Reply(nil), ParentUserId:"", Files:[]slack.File(nil), Upload:false, Comment:(*slack.Comment)(nil), ItemType:"", ReplyTo:0, Team:"T216B1U03", Reactions:[]slack.ItemReaction(nil), ResponseType:"", ReplaceOriginal:false, DeleteOriginal:false, Blocks:slack.Blocks{BlockSet:[]slack.Block(nil)}}, SubMessage:(*slack.Msg)(nil), PreviousMessage:(*slack.Msg)(nil)}
```

**Comment by qaisjp** [comment]
More context.

Running the following code (using the bot token)

```
func main() {
	api := slack.New("xoxb-xxxx-yyyy-zzzz")
	api.SendMessage("C4Z7JEUC8", slack.MsgOptionText("kek", true))
}
```

Sends this message:

![image](https://user-images.githubusercontent.com/923242/76855581-55d2c880-6849-11ea-9d4c-1381b007592b.png)

Logs this:

```
Comparing (b.rtm!=nil)[true] && ev.Username["Chat Bridge"] == b.si.User.Name["chat-bridge"]

slack b.si.User.ID "UF3N0SGJU" and ev.BotID "BF5S2HSTG" and ev.User ""

Slack event: &slack.MessageEvent{Msg:slack.Msg{ClientMsgID:"", Type:"message", Channel:"C4Z7JEUC8", User:"", Text:"kek", Timestamp:"1584446901.006900", ThreadTimestamp:"", IsStarred:false, PinnedTo:[]string(nil), Attachments:[]slack.Attachment(nil), Edited:(*slack.Edited)(nil), LastRead:"", Subscribed:false, UnreadCount:0, SubType:"bot_message", Hidden:false, DeletedTimestamp:"", EventTimestamp:"1584446901.006900", BotID:"BF5S2HSTG", Username:"Chat Bridge", Icons:(*slack.Icon)(nil), Inviter:"", Topic:"", Purpose:"", Name:"", OldName:"", Members:[]string(nil), ReplyCount:0, Replies:[]slack.Reply(nil), ParentUserId:"", Files:[]slack.File(nil), Upload:false, Comment:(*slack.Comment)(nil), ItemType:"", ReplyTo:0, Team:"T216B1U03", Reactions:[]slack.ItemReaction(nil), ResponseType:"", ReplaceOriginal:false, DeleteOriginal:false, Blocks:slack.Blocks{BlockSet:[]slack.Block(nil)}}, SubMessage:(*slack.Msg)(nil), PreviousMessage:(*slack.Msg)(nil)}
```

More interestingly, my app appears in the apps list, which I can PM:

![image](https://user-images.githubusercontent.com/923242/76855752-b235e800-6849-11ea-948b-6e584843b230.png)


The old version of the same app, also appears in the contact list:

![image](https://user-images.githubusercontent.com/923242/76855780-c4178b00-6849-11ea-9011-6d8c355bc1a1.png)

Which is apparently a deactivated account:

![image](https://user-images.githubusercontent.com/923242/76855796-cc6fc600-6849-11ea-923b-afb936df8365.png)

And the app is apparently disabled:

![image](https://user-images.githubusercontent.com/923242/76855811-d5609780-6849-11ea-8802-c62fb60b996b.png)

Clicking "How does chat-bridge work?" shows this

![image](https://user-images.githubusercontent.com/923242/76855835-dee9ff80-6849-11ea-9f4e-65cbf96c85bc.png)

Clicking "Settings" or "View in App Directory" will both link to https://multitheftauto.slack.com/apps/AF4ACB3V2-chat-bridge, which says "There's been a glitch":

![image](https://user-images.githubusercontent.com/923242/76855890-ffb25500-6849-11ea-94eb-ec1a03d6ae6b.png)

On Slack, "Your Apps", my `Chat Bridge` app (at https://api.slack.com/apps/AF3RWLXBK/app-home) shows this:

![image](https://user-images.githubusercontent.com/923242/76855971-32f4e400-684a-11ea-8636-e42ad2314189.png)

**Observe "legacy Bot User"**, which is possibly why I am running into issues.

Note that inviting the bridge shows "Bridge" and not "chat-bridge". Originally (when I first set the bridge up) it would say `@chat-bridge`:

![image](https://user-images.githubusercontent.com/923242/76856047-5f106500-684a-11ea-9aac-8dc8b231ceca.png)

On another Slack workspace (with a different _app_ so a different token, but the same configuration), `@chat-bridge` works fine - https://api.slack.com/apps/A9C11GZ7H/app-home?:

![image](https://user-images.githubusercontent.com/923242/76856422-22913900-684b-11ea-9e8d-9232f0918765.png)

On the `OAuth & Permissions` tab, both of them say this:

![image](https://user-images.githubusercontent.com/923242/76856447-3341af00-684b-11ea-8fa5-1e03efe98b6d.png)

----

Do you think I should just delete the app and start over? Or should I keep this around so that we can be legacy bots work with matterbridge? Should we support legacy bots?

**Comment by qaisjp** [comment]
Aha! Above, in the "installed bots" (not the developer interface) screen, under the "Bot User" heading, the usernames are different.

**Comment by qaisjp** [comment]
I'll try and see if we can make this use the ID only, and not compare usernames.

It still won't explain why it works fine on Linux (with an older version of matterbridge, which has a different copy of the library). But /shrug, it should solve the issue.

**Comment by qaisjp** [comment]
![image](https://user-images.githubusercontent.com/923242/76877951-06ea5a80-686c-11ea-90fc-0b893be7f208.png)

I can confirm this happens on 6b4b19194ec40c79569bc692a6557e78f338ded9 on my production instance (just upgraded). 

After cherry-picking #1042 the messages are no longer duplicated.

**Comment by The-Sorce** [comment]
I also experienced the same issue with a simple Slack<->Discord bridge setup.

Can confirm the issue but I have not tested the fix.

**Comment by qaisjp** [comment]
the core issue ALL ALONG is that something is wrong with attachments, lol

![image](https://user-images.githubusercontent.com/923242/77131202-b5e88b00-6a52-11ea-94c5-3f816d525a4b.png)

**Comment by qaisjp** [comment]
The fix in #1042 went in a completely incorrect direction.

#1046 does it properly.
